spin/graph.go

```diff
@@ -0,0 +1,52 @@
+package spin
```

Comment:
All the interesting code in this PR is in the `spin` package. Since this is a massive PR due to all the logging refactoring, I'd recommend starting here!